### DISCUSSION TITLE
Error converting XML to ER7 when adding source transformer

### DISCUSSION THREAD

**Discussion by nickmillett** [nextgenhealthcare/connect#5139] [Q&A] [answered]
My team has been using Mirth 3.2.1 for many years, and are now in the process of upgrading to 3.12. I am seeing some unexpected behavior with the source transformer JavaScript.

I have created a brand new channel with default TCP listener. This works fine and I am able to send and process messages. However, when I add any source transformer with JavaScript (even just a single `return;`) I get the following error when receiving messages:

`ER7Serializer error
ERROR MESSAGE: Error converting XML to ER7
org.xml.sax.SAXParseException; Premature end of file.
	at org.apache.xerces.util.ErrorHandlerWrapper.createSAXParseException(Unknown Source)
	at org.apache.xerces.util.ErrorHandlerWrapper.fatalError(Unknown Source)
	at org.apache.xerces.impl.XMLErrorReporter.reportError(Unknown Source)
	at org.apache.xerces.impl.XMLErrorReporter.reportError(Unknown Source)
	at org.apache.xerces.impl.XMLErrorReporter.reportError(Unknown Source)
	at org.apache.xerces.impl.XMLVersionDetector.determineDocVersion(Unknown Source)
	at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source)
	at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source)
	at org.apache.xerces.parsers.XMLParser.parse(Unknown Source)
	at org.apache.xerces.parsers.AbstractSAXParser.parse(Unknown Source)
	at com.mirth.connect.plugins.datatypes.hl7v2.ER7Serializer.fromXML(ER7Serializer.java:291)
	at com.mirth.connect.donkey.server.channel.FilterTransformerExecutor.processConnectorMessage(FilterTransformerExecutor.java:122)
	at com.mirth.connect.donkey.server.channel.Channel.process(Channel.java:1661)
	at com.mirth.connect.donkey.server.channel.Channel.dispatchRawMessage(Channel.java:1288)
	at com.mirth.connect.donkey.server.channel.SourceConnector.dispatchRawMessage(SourceConnector.java:192)
	at com.mirth.connect.donkey.server.channel.SourceConnector.dispatchRawMessage(SourceConnector.java:170)
	at com.mirth.connect.connectors.tcp.TcpReceiver$TcpReader.call(TcpReceiver.java:680)
	at com.mirth.connect.connectors.tcp.TcpReceiver$TcpReader.call(TcpReceiver.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)`

Through using some log statements, I have identified that the `msg` variable is empty in the source transformer. If I add the following as the first line in the transformer, it works:

`msg = SerializerFactory.getSerializer('HL7V2').toXML(connectorMessage.getRawData());`

Why is this the case? When we were using Mirth 3.2.1 the `msg` variable was always available without having to define it explicitly like this. I am unsure if adding this line is now necessary, or if this is indicative of something else missing in the channel configuration. Thanks for your help!

**Comment by pacmano1** [comment]
If that is a plain TCP listener (not MLLP) have you and the sender agreed to what constitutes the end of message and have you configured that in mirth on the listener?   

If it is TCP listener is the data type set to raw?  (since you are serializing it)

If it is MLLP, have you verified the sender is ending a message as expected?

That is the little wrench by the drop down where you select the Mode right under "TCP Listener Settings"

**Reply by nickmillett** [reply]
Thanks for the suggestions. I took a look at that section of my new channel and our previously existing channels. We have always used MLLP with the default start and end of message bytes, which has worked for us in version 3.2.1. 

I have an instance of 3.2.1 running and have sent the same message to 3.2.1 and 3.12, with the same brand new channels created in each, with the same settings and with a JavaScript source transformer just containing `return;`. In version 3.2.1 the message processes and gets a status of SENT, but in 3.12 it gets the error above unless I define the `msg` variable in the source transformer...

**Reply by pacmano1** [reply]
Maybe post a bare minimum export of your channel that replicates the behavior here and a message that fails.   3.2.1 is ancient I don't remember much if it.  3.12 channel I mean, not the 3.1.2 export.

**Reply by nickmillett** [reply]
Sure, I have attached a ZIP package which contains exports of identical channels I created separately in both Mirth 3.12 and 3.2.1, the sample HL7 message, and screenshots of the results.

[mirth_312_testing.zip](https://github.com/nextgenhealthcare/connect/files/8496356/mirth_312_testing.zip)

**Reply by pacmano1** [reply]
I imported your channel to 3.12, sent in your message, no problems.  Is there any device in front of the port listening?  IDS, IPS? stunnel? etc?   

And how are you sending the message to mirth?   If you send to mirth from itself do you see this?

**Reply by nickmillett** [reply]
Thanks for checking. Yes, we have some surrounding software which could be affecting the behavior - namely, we are running Mirth inside Docker containers. This was the same when using Mirth 3.2.1, but I did a test of the channel with Mirth 3.12 in an instance running outside the container and it worked. I'm now investigating any Docker configuration that could be resulting in the parse exception in Mirth.

**Reply by pacmano1** [reply]
LOL, people razz me all the time when I tell them containers drive me up the wall,   I hereby thank you for a post I can point to, lol.  Do post an update when you figure it out!

**Reply by siddharthsharma9258** [reply]
I think that docker is removing CR LF characters from the raw message before relaying it onto mirth.

**Reply by mirkoesins** [reply]
I am facing the same issue. The only differences are that I am coming from version 3.10 and not using docker. Do you found reason why it does not work? I tested the channel you posted here too but the outcome is the same to my own channels. Until version 3.11 Mirth worked properly for me.

**Reply by nickmillett** [reply]
In our case, we determined the issue was caused by some JARs that were included in our custom-lib directory. Specifically, we had Saxon-HE-9.6.0-7.jar and xalan-2.7.2.jar in that directory. We found that updating the Saxon JAR to 11.3 and completely removing the xalan JAR fixed the "Error converting XML to ER7" problem and also allowed our custom transformers to continue to work (that's why we had to keep the Saxon JAR).

It turns out the issue wasn't related to Docker specifically, but rather in my earlier non-Docker test it was using an empty custom-lib directory.

**Reply by mirkoesins** [reply]
Worked out for me too. Thanks for your reply!

**Reply by tonygermano** [reply]
I would have been very surprised if docker was altering the data coming across a TCP connection.

**Comment by pacmano1** [comment, marked as answer]
OP answered his own problem, adding this to mark as answered
```
It turns out the issue wasn't related to Docker specifically, but rather in my earlier non-Docker 
test it was using an empty custom-lib directory.   Problem was due to jars in custom-lib.
```
